### PR TITLE
fix: bound oversized user-message rendering (#3797)

### LINKED ISSUE #3797: [Bug]: A ~3900-line pasted message degrades the whole web UI and times out WebSocket RPCs
### Summary

Adding a single very large user message to a session (about 3921 lines of pasted console log output) degrades that session and the entire web UI. The session slows down and stops being able to finish its tasks, the whole Kandev screen becomes sluggish while that session is selected, and switching to another session takes a very long time. While this is happening, WebSocket RPCs time out, and a "Couldn't start a session / WebSocket request timed out: session.launch" banner appears roughly once a minute.

### Affected area

Web UI

### Kandev version

v0.94.0 (commit bf819a02, built 2026-09-09T16:46:24Z)

### Install mode

Other: published binary installed via npm/Homebrew, run in unified mode.

### Environment

- OS: Linux (kernel 6.17.0-1032-oem)
- Architecture: amd64
- Browser: Chrome
- Install: published Kandev binary, unified mode, served on 127.0.0.1
- Database: SQLite at `~/.kandev/data/kandev.db`
- Docker/Kubernetes: not used

### Kandev configuration

The affected task has since been deleted, so its configuration could not be read back:

- Executor type/profile: not recorded
- Agent profile: not recorded
- Repository source: not recorded
- Relevant `KANDEV_*` env vars: not recorded

### Task/session context

- Affected task ID: `42b88450-a174-48fd-85b6-586f45114809` (local UUID; the task has since been deleted, so the oversized message is no longer in the database and its exact byte size could not be measured after the fact)
- Affected session ID: `8ce2f055-d191-4671-b1f2-540a09777f2e`
- Timeframe of the degradation: 2026-09-18 06:32 to 07:57 UTC (all timestamps in this report are normalized to UTC; backend logs are stamped +0300)
- The largest user message still present in the database is 12,092 bytes / 157 lines, so the reported payload was substantially larger than anything that survives for comparison.

### Steps to reproduce

1. Open a running task session in the web UI.
2. Paste a very large block of text into the message composer as a comment for the agent. In this case it was roughly 3921 lines of console log output.
3. Send the message.
4. Keep the session selected and interact with the UI.
5. Switch to a different session.

### Reproducibility

Always

### Clean-state check

Not tried

### Expected behavior

A large pasted message is accepted and stored without degrading interactive performance. The conversation view stays responsive, the agent continues to make progress on its tasks, unrelated sessions are unaffected, and switching sessions remains fast. If a message is too large to handle, Kandev should bound it explicitly (truncate, collapse, or attach it as a file) rather than degrading the whole client.

### Actual behavior

After the large message was added:

- The session became very slow and could not finish its tasks.
- The entire Kandev screen was sluggish while that session was selected, not just the conversation panel.
- Switching to another session took a very long time.
- A "Couldn't start a session / WebSocket request timed out: session.launch" error banner appeared roughly once per minute.

### Logs and artifacts

The oversized message itself is not reproduced here (it was 3921 lines of console output, and the task has since been deleted).

A diagnostic bundle captured from the affected instance shows 64 `WebSocket request timed out: <method>` client errors, with the vast majority concentrated on the day of the incident:

```text
Timeouts per day:
  2026-09-16:  7
  2026-09-17:  4
  2026-09-18: 53
```

They are concentrated on the affected task, and they are not specific to one RPC method:

```text
task 42b88450 (affected)  34 timeouts  06:32:25Z - 07:55:30Z
    session.git.commits        12
    session.cumulative_diff    10
    message.queue.get           8
    user_shell.list             2
    task.plan.revisions.list    1
    message.add                 1

other tasks, same instance, same window
    task e4e9916b   8 timeouts  06:31:35Z
    task 08db06f5   7 timeouts  07:04:12Z
```

Representative client error:

```text
Failed to send message:
Error: WebSocket request timed out: message.add
    at /assets/task-status-summary-*.js
```

Two things stand out in this data:

1. The timeouts span six unrelated RPC methods rather than one, which points at a shared bottleneck (the WebSocket connection or the client's main thread) rather than a fault in any single handler.
2. Other tasks open in the same browser also timed out during the same window, which matches the report that the whole UI, not just the affected session, became slow.

Backend logs from the same window **do** contain correlating errors, two of which name the affected task and session directly:

```text
06:38:23Z  ERROR  prompt completed with error
           error: agent stream disconnected: read tcp 127.0.0.1:51930->127.0.0.1:41004:
                  use of closed network connection
06:38:23Z  ERROR  initial prompt failed  (same execution)

07:54:34Z  ERROR  prompt completed with error
           error: agent stream disconnected: read tcp 127.0.0.1:52548->127.0.0.1:41004:
                  use of closed network connection
07:54:34Z  ERROR  prompt failed
           component: orchestrator
           task_id:   42b88450-a174-48fd-85b6-586f45114809
           session_id: 8ce2f055-d191-4671-b1f2-540a09777f2e

07:56:42Z  ERROR  prompt failed   (same task and session, same error)
```

The backend also recorded the forwarded client toast at the moment of the `message.add` timeout:

```text
07:55:30Z  ERROR  frontend error toast
           title:       Message send status unknown
           description: The connection dropped or timed out. Refresh the task to
                        confirm whether it went through.
           task_id:     42b88450-a174-48fd-85b6-586f45114809
```

So the agent stream to agentctl on port 41004 was torn down (`use of closed network connection`) and prompts on the affected session failed repeatedly. That is consistent with the session being unable to finish its tasks, and it suggests the fault is not purely client-side rendering.

One discrepancy worth flagging: the error banner in the UI reads `session.launch`, but `session.launch` does not appear anywhere in the captured client timeout breakdown above. The banner may be raised by a retry path distinct from the RPCs that were logged, or the captured window may simply not overlap the banner occurrences.

### Last known good version

Unknown. This is the first time a message of this size was added.

### Extra context

Possibly related prior work, though neither covers an oversized conversation message:

- #2893 — kanban board hangs with ~440 tasks in one column; server answers in 50 ms, so the cost is client-side rendering. Same shape of problem (client-side rendering cost, not backend latency) in a different view.
- #3663 — `message.queue.add` submit was fire-and-forget with a default 5s WebSocket timeout, and failures surfaced as "connection lost".

Two areas that may be worth checking:

- Whether the conversation view renders a very large message in full (Markdown parsing, syntax highlighting, virtualization) on every render, which would explain UI-wide slowness and slow session switching.
- Whether a large message is re-sent over the WebSocket on session load or subscription, which would explain unrelated RPCs on the same connection timing out.
- Why the agent stream to agentctl closed (`use of closed network connection` on port 41004) while the oversized message was in the session. If a large prompt payload can tear down or stall that stream, the failed prompts and the client-side slowness may share one cause rather than being two separate faults.

### Before submitting

- [x] I can reproduce this using the steps above.
- [x] I removed secrets, tokens, private repository URLs, private prompts, and customer data.
- [x] I searched existing issues and did not find a duplicate.

### PR DESCRIPTION
**Pull request by carlosflorencio**
> [!TIP]
> **PR walkthrough:** [Open the visual walkthrough](https://walkthrough.kandev.ai/pr/3801/b841a52a20c7.html)

Oversized user messages could make the browser render thousands of Markdown nodes and lose transcript responsiveness. This change bounds every affected preview while keeping the complete message available for download, storage, editing, queue operations, and agent input.

## Important Changes

- Bound transcript, raw, workflow-instruction, queued, and pinned previews to 200 logical lines or 16,000 UTF-16 code units before Markdown rendering.
- Preserve complete message downloads, including text on both sides of a disclosed workflow instruction block.
- Add desktop and phone regression flows with the 3,921-line, approximately 328 KB incident-sized log, in-app task navigation, reload coverage, exact download assertions, and successful follow-up persistence checks.
- Update localized copy, public usage documentation, requirements, system design, and the implementation work order.
- Correct the pre-existing duplicate queued-session acceptance ID so specification lint and commit hooks pass.

## Validation

- Focused Vitest suite: 7 files, 141 tests passed.
- `cd apps/web && pnpm run typecheck`
- `cd apps/web && pnpm run lint` (zero warnings)
- `cd apps/web && pnpm run i18n:check`
- `cd apps/web && pnpm run build:vite`
- Desktop managed E2E: 13 tests passed for oversized-message and last-prompt flows.
- Mobile managed E2E: 2 tests passed for oversized-message and last-prompt flows.
- Standalone desktop oversized-message E2E: 2 tests passed.
- Standalone mobile oversized-message E2E: 1 test passed.
- `node --test scripts/validate-public-docs.test.mjs && node scripts/validate-public-docs.mjs`
- `python3 scripts/list-docs.py validate`
- `python3 scripts/lint-spec-files.test.py`
- `python3 scripts/lint-spec-files.py --all`
- `git diff --check`
- Fresh synthetic desktop and mobile PR captures from the managed E2E runner were inspected, compressed, and validated against their asset manifest.

## Checklist

- [ ] If I do not have repository write access and this is a large architectural change, I discussed the direction in a linked issue before opening this PR.
- [ ] This PR contains one logical change; unrelated work is split into separate PRs.
- [ ] I have performed a self-review of my code.
- [ ] I have manually tested my changes and they work as expected.
- [ ] My changes have tests that cover the new functionality and edge cases.
- [ ] If my change touches UI files (`apps/web/`), I have added or updated Playwright e2e tests in `apps/web/e2e/` and verified them with `make test-e2e`.
- [ ] I checked whether this affects public docs in `docs/public/**` and updated them or noted why no docs change is needed.

Closes #3797

<a href="https://cubic.dev/pr/kdlbs/kandev/pull/3801?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Screenshots

Desktop bounded oversized-message preview

Mobile bounded oversized-message preview

### CODE REVIEW

File: apps/web/components/task/chat/messages/bounded-message-preview.tsx [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+          <span>{t("task:messagePreviewShortened")}</span>
+          <Button
+            type="button"
+            variant="link"
+            onClick={download}
+            aria-label={t("task:downloadFullMessage")}
+            data-testid={`${testId}-download`}
+            className={cn(
+              "shrink-0 px-0 text-xs",
+              isFinePointer ? "h-7" : "h-11 w-full justify-start",
```
[the remark this answers is not in the document]
Comment (carlosflorencio) [reply apps/web/components/task/chat/messages/bounded-message-preview.tsx]: Addressed in commit b841a52a20c7a4c2eeaff246aa7323b3a83b4124. The download action now keeps the 44px mobile target whenever `isMobile` is true, even with a fine pointer. Added component coverage and a 500px fine-pointer Playwright regression; the focused desktop and mobile oversized-message suites pass.
